PostSharp.Community.StructuralEquality.Weaver/EqualsInjection.cs

```diff
@@ -0,0 +1,408 @@
+
+            var typeTypeDef = project.Module.Cache.GetType( typeof(Type) ).GetTypeDefinition();
+            this.getTypeFromHandleMethod = project.Module.FindMethod( typeTypeDef, "GetTypeFromHandle" );
+
+            this.equatableInterface = project.Module.Cache.GetType( typeof(IEquatable<>) );
+        }
+
+        public void AddEqualsTo( TypeDefDeclaration enhancedType, StructuralEqualityAttribute config,
+            ISet<FieldDefDeclaration> ignoredFields )
+        {
+            var typedEqualsMethod = this.InjectEqualsType( enhancedType, config, ignoredFields );
+            this.InjectEqualsObject( enhancedType, config, typedEqualsMethod );
+
+            this.AddEquatable( enhancedType );
+        }
+
+        private void AddEquatable( TypeDefDeclaration enhancedType )
+        {
+            var genericInstance = this.equatableInterface.GetTypeDefinition().GetGenericInstance( new GenericMap(
+                enhancedType.Module,
+                new[] {enhancedType.GetCanonicalGenericInstance()} ) );
+            enhancedType.InterfaceImplementations.Add( genericInstance.TranslateType( enhancedType.Module ) );
+        }
+
+        private MethodDefDeclaration InjectEqualsType( TypeDefDeclaration enhancedType,
+            StructuralEqualityAttribute config, ICollection<FieldDefDeclaration> ignoredFields )
+        {
+            // public virtual bool Equals( Typed other )
+            var equalsDeclaration = new MethodDefDeclaration
+            {
+                Name = "Equals",
+                Attributes = MethodAttributes.Public | MethodAttributes.HideBySig | MethodAttributes.Virtual,
+                CallingConvention = CallingConvention.HasThis
+            };
+            enhancedType.Methods.Add( equalsDeclaration );
+            equalsDeclaration.Parameters.Add( new ParameterDeclaration( 0, "other",
+                enhancedType.GetCanonicalGenericInstance() ) );
+            equalsDeclaration.ReturnParameter = ParameterDeclaration.CreateReturnParameter( this.booleanType );
+            CompilerGeneratedAttributeHelper.AddCompilerGeneratedAttribute( equalsDeclaration );
+
+            using ( var writer = InstructionWriter.GetInstance() )
+            {
+                var methodBody = MethodBodyCreator.CreateModifiableMethodBody( writer, equalsDeclaration );
+                InstructionSequence instructionSequence = methodBody.PrincipalBlock.AddInstructionSequence();
+                instructionSequence.Comment = "Equals(Typed other)";
+                writer.AttachInstructionSequence( instructionSequence );
+
+                this.InjectReferenceEquals( writer, methodBody, enhancedType );
+                // Writer is either attached to the same sequence (value types) or to a new one which should check the structure.
+
+                // return base.Equals(other) && this.field1 == other.field1 && ...;
+                if ( !config.IgnoreBaseClass && !enhancedType.IsValueType() )

```

Comment:
I've had trouble sometimes with `IsValueType` throwing exceptions and I tend to use `IsValueTypeSafe` instead.
```suggestion
                if ( !config.IgnoreBaseClass && !enhancedType.IsValueTypeSafe() == true )

```